[TAN-3358] Upgrade Rails to version 7.1.5.1 (Part 1)

adessy opened the pull request:
The upgrade will be rolled out in two steps/releases. Part 2 is here: https://github.com/CitizenLabDotCo/citizenlab/pull/10140.

# Changelog
## Technical
- Upgrade Rails from 7.0.8.7 to 7.1.5.1.
- Upgrade Ruby to version 3.3.4.

adessy committed:
[TAN-3358] Bump active_model_serializers to 0.10.15

adessy committed:
[TAN-3358] Upgrade Ruby to version 3.3.4

Upgrade Ruby to resolve the following bug encountered when upgrading Rails to version 7.1.5.1:
```
ArgumentError: comparison of String with nil failed (ArgumentError)
    msg = " #{RUBY_VERSION < SINCE[gem] ? "will no longer be" : "is not"} part of the default gems since Ruby #{SINCE[gem]}."
```

Refer to https://bugs.ruby-lang.org/issues/20450 for more details.

adessy committed:
[TAN-3358] Upgrade Rails to version 7.1.5.1

adessy committed:
[TAN-3358] Update the locked version of Bundler to v2.5.11

This is the version that is included in the `ruby:3.3.4-slim` image.

adessy commented on back/config/initializers/new_framework_defaults_7_1.rb:
```diff
@@ -0,0 +1,270 @@
+#   "X-Content-Type-Options" => "nosniff",
+#   "X-Permitted-Cross-Domain-Policies" => "none",
+#   "Referrer-Policy" => "strict-origin-when-cross-origin"
+# }
```
@kogre okay to drop support for Internet Explorer? I checked Matomo, and it does not appear the browser report.

kogre replied:
Yes, we dropped support along time ago

adessy committed:
[TAN-3358] Disable `allow_deprecated_parameters_hash_equality`

adessy committed:
[TAN-3358] Set up the default behavior of Rails 7.1+ for AR encryption

adessy committed:
[TAN-3358] Remove the default X-Download-Options headers

adessy committed:
[TAN-3358] Configures SQLite with a strict strings mode

adessy committed:
[TAN-3358] Disable deprecated singular associations names

adessy committed:
[TAN-3358] Raise `ArgumentError` for invalid `expires_at` or `expires_in` with `Rails.cache`.

adessy committed:
[TAN-3358] Format tags using the SQLCommenter format

adessy committed:
[TAN-3358] Set the maximum size for Rails log files

adessy committed:
[TAN-3358] Raise an exception when assigning to `attr_readonly` attributes

adessy committed:
[TAN-3358] Enable some optimizations

adessy committed:
[TAN-3358] Disable automatic column serialization into YAML

adessy committed:
[TAN-3358] Enable before_committed! callbacks on all enrolled records in a transaction

adessy committed:
[TAN-3358] Configure Action View to use HTML5 standards-compliant sanitizers

The application does not use Action Text.

adessy committed:
[TAN-3358] Configure the log level used by the DebugExceptions middleware

adessy committed:
[TAN-3358] Configure the test helpers to use HTML5 parsers

adessy committed:
[TAN-3358] Configure `return` behavior in `transaction` blocks

adessy committed:
[TAN-3358] Use 7.1 default behaviour for after_commit callbacks

adessy committed:
[TAN-3358] Use Rails 7.1 default serializer for `MessageEncryptor` and `MessageVerifier`

adessy committed:
[TAN-3358] Run `after_commit` and `after_*_commit` callbacks in the order they are defined in a model.

adessy committed:
[TAN-3358] Use Rails 7.1 default behavior for `has_secure_token` declarations

adessy committed:
[TAN-3358] Remove call to deprecatedv`Rails.application.secrets`

> DEPRECATION WARNING: `Rails.application.secrets` is deprecated in favor of `Rails.application.credentials` and will be removed in Rails 7.2.

But it's simpler to access `secret_key_base` using `Rails.application.secret_key_base`.

adessy committed:
[TAN-3358] Set Rails 7.1 default for `add_autoload_paths_to_load_path`

kogre commented on back/config/initializers/filter_parameter_logging.rb:
```diff
@@ -2,11 +2,11 @@
+# See the ActiveSupport::ParameterFilter documentation for supported notations and behaviors.
 Rails.application.config.filter_parameters += %i[
-  password current_password passw secret token _key crypt salt certificate otp ssn import_ideas.pdf import_ideas.xlsx
+  passw secret token _key crypt salt certificate otp ssn import_ideas.pdf import_ideas.xlsx
```
Deliberate change to remove `password` and `current_password` from this list?

adessy replied:
Yep, according to the documentation, it's supposed to use partial matching. So, `password` and `current_password` are covered by `passw`.

kogre reviewed (approved):
I didn't study all the required changes, so I assume you're doing all of these with a good reason :)

Could be good to highlight some changes in the CDM? Thinking of the change to deep_merge, the change to parameters to_h and Rails.env.local?

adessy committed:
Merge branch 'master' into TAN-3358/upgrade-to-rails-7.1

# Conflicts:
#	back/Gemfile.lock

adessy committed:
Merge branch 'master' into TAN-3358/upgrade-to-rails-7.1

# Conflicts:
#	back/Gemfile.lock
